### PR TITLE
Add dynamic sidecar tasks

### PR DESCRIPTION
**Pull request by jeevb**
# Adds a new task type: `dynamic_sidecar_task`, that allows for running dynamic tasks with custom pod specs.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
The new class `SdkDynamicSidecarTask` is a hybrid of the dynamic and sidecar tasks. The core logic in `SdkDynamicTask` was moved into a mixin (`SdkDynamicTaskMixin`) that was then used as a base in `SdkDynamicTask` and `SdkDynamicSidecarTask` respectively. `SdkDynamicSidecarTask` allows us to use the same mechanisms for specifying custom pod specs in `sidecar_task`s for `dynamic_task`s as well!

## Tracking Issue
https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
NA

### CODE REVIEW

File: flytekit/common/tasks/sdk_dynamic.py [hunk 1]
```diff
@@ -51,65 +51,8 @@ def _append_node(generated_files, node, nodes, sub_task_node):
...
-        :param int max_concurrency:
-        :param dict[Text, Text] environment:
-        :param dict[Text, T] custom:
-        """
-        super(SdkDynamicTask, self).__init__(
-            task_function, task_type, discovery_version, retries, interruptible, deprecated,
-            storage_request, cpu_request, gpu_request, memory_request, storage_limit,
-            cpu_limit, gpu_limit, memory_limit, discoverable, timeout, environment, custom)
-
+class SdkDynamicTaskMixin:
```
Comment (wild-endeavor) [inline flytekit/common/tasks/sdk_dynamic.py:54]: let `(object)` here for completeness.

File: flytekit/common/tasks/sdk_dynamic.py [hunk 2]
```diff
@@ -305,3 +248,65 @@ def execute(self, context, inputs):
             })
 
             return generated_files
+
+
+class SdkDynamicTask(_six.with_metaclass(_sdk_bases.ExtendedSdkType, SdkDynamicTaskMixin, _sdk_runnable.SdkRunnableTask)):
```
Comment (wild-endeavor) [inline flytekit/common/tasks/sdk_dynamic.py:253]: we can stop using six - in the refactor that we're doing we're clearing out a lot of the python2 stuff.  feel free to just `metaclass= _sdk_bases.ExtendedSdkType`

File: flytekit/common/tasks/sidecar_task.py [hunk 3]
```diff
@@ -138,3 +139,52 @@ def reconcile_partial_pod_spec_and_task(self,
         ).to_flyte_idl()
 
         self.assign_custom_and_return(_MessageToDict(sidecar_job_plugin))
+
+
+class SdkDynamicSidecarTask(_six.with_metaclass(_sdk_bases.ExtendedSdkType, _sdk_dynamic.SdkDynamicTaskMixin, SdkSidecarTask)):
```
Comment (wild-endeavor) [inline flytekit/common/tasks/sidecar_task.py:144]: same here

**Comment by kumare3** [comment]
@jeevb this is actually pretty cool and explains that our interface for task construction is incorrect.
I think it should be
```python
@task(type=Sidecar(pod_spec=...), dynamic=True, ....)
def func(...):
```

**Comment by jeevb** [comment]
@kumare3: I like where you are going with that, but it isn't clear to me how it will play with other task types like `@spark_task` though. It does make sense for `@python_task` and `@sidecar_task`.

**Comment by kumare3** [comment]
@jeevb it will be the same
```python
@task(type=Spark(spark_conf=...), resources=...)
```
But also we dont have to use it for all task types, some task types may have dedicated decorators (I really want to avoid that, to simplify the design language)

**Comment by jeevb** [comment]
Ah I see. That makes sense!

**Comment by wild-endeavor** [comment]
let me run this through our other tests tomorrow morning before merging.

File: flytekit/__init__.py [hunk 4]
```diff
@@ -2,4 +2,4 @@
 
 import flytekit.plugins
 
-__version__ = '0.11.4'
+__version__ = '0.11.5'
```
Comment (wild-endeavor) [inline flytekit/__init__.py:5]:
```suggestion
__version__ = '0.11.6'
```

**Pull request merged**